src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,108 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_CAPACITY = 1.5;
+    private T[] array;
+    private int size;
+
+    public ArrayList() {
+        array = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public void growthArray() {
+        T[] arrayCopy = (T[]) new Object[(int) (array.length * GROWTH_CAPACITY)];
+        System.arraycopy(array, 0, arrayCopy, 0, array.length);
+        array = arrayCopy;
+
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == array.length) {
+            growthArray();
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index "
+                    + index + " is not valid for size " + size + ".");
+        }
+        if (size == array.length) {
+            growthArray();
+        }
 
+        if (index < size) {
+            System.arraycopy(array, index, array, index + 1, size() - index);
+            array[index] = value;
+            size++;
+        }
+        if (index == size) {
+            add(value);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public T remove(int index) {
+        Object removeElement;
+        checkIndex(index);
+        removeElement = array[index];
+        System.arraycopy(array, index + 1, array, index, size - index - 1);
+        size--;
+        return (T) removeElement;
     }
 
     @Override
-    public void set(T value, int index) {
-
+    public T remove(T element) {
+        for (int i = 0; i < size(); i++) {
+            if ((array[i] == null && element == null)
+                    || element != null && Objects.equals(element, array[i])) {
```

Comment:
```suggestion
            if ((array[i] == element)
                    || element != null && element.equals(array[i])) {
```
<img width="208" alt="Screen Shot 2022-12-29 at 12 21 01 PM" src="https://user-images.githubusercontent.com/90692308/209945818-436f4c2c-16bc-4467-bded-147e26d92364.png">